extensions/amp-form/0.1/form-verifiers.js

```diff
@@ -0,0 +1,326 @@
+      return new VerificationGroup(elements);
+    });
+  } else {
+    throw new Error('The amp-form verification config should ' +
+        'be put in a <script> tag with type="application/json"');
+  }
+}
+
+/**
+ * An interface for a form verifier. Implementations could check for duplicate
+ * usernames on a remote server, check against an in-memory cache to verify
+ * data in ways not possible with standard form validation, or check
+ * values against sets of data too large to fit in browser memory
+ * e.g. ensuring zip codes match with cities.
+ * @visibleForTesting
+ */
+export class FormVerifier {
+  /**
+   * @param {!HTMLFormElement} form
+   * @param {!Array<!VerificationGroup>} config
+   * @param {function():Promise<!../../../src/service/xhr-impl.FetchResponse>} xhr
+   */
+  constructor(form, config, xhr) {
+    /** @protected @const */
+    this.form_ = form;
+
+    /** @protected @const {!Array<!VerificationGroup>} */
+    this.groups_ = config;
+
+    /** @protected @const*/
+    this.doXhr_ = xhr;
+  }
+
+  /**
+   * Called when the user has modified the value in the input,
+   * e.g. the input's 'input' event
+   * @param {!Element} unusedInput
+   */
+  onMutate(unusedInput) {}
+
+  /**
+   * Called when the user has fully set a value to be verified,
+   * e.g. the input's 'change' event
+   * @param {!Element} unusedInput
+   * @param {!function()} unusedAfterVerify
+   */
+  onCommit(unusedInput, unusedAfterVerify) {}
+}
+
+/**
+ * A no-op verifier.
+ * @visibleForTesting
+ */
+export class DefaultVerifier extends FormVerifier { }
+
+/**
+ * A verifier that verifies values via an XHR
+ * @visibleForTesting
+ */
+export class AsyncVerifier extends FormVerifier {
+  /** @override */
+  onMutate(input) {
+    const group = this.getGroup_(input);
+    if (group) {
+      group.setDirty(true);
+      group.clearErrors();
+    }
+  }
+
+  /** @override */
+  onCommit(input, afterVerify) {
+    if (this.isVerificationElement_(input)) {
+      this.maybeVerify_(afterVerify);
+    }
+  }
+
+  /**
+   * Sends the verify request if any group is ready to verify.
+   * @param {!function()} afterVerify
+   */
+  maybeVerify_(afterVerify) {
+    if (this.shouldVerify_()) {
+      this.doXhr_().then(() => {
+        return [];
+      }, error => {
+        return getResponseErrorData(/** @type {!Error} */ (error));
+      })
+      .then(errors => this.verify_(errors))
+      .then(() => afterVerify());
+    }
+  }
+
+  /**
+   * Set errors on elements that failed verification, and clear any
+   * verification state for elements that passed verification.
+   * @param {!Array<!VerificationErrorDef>} errors
+   */
+  verify_(errors) {
+    const errorElements = [];
+
+    // Set the error message on each element that caused an error.
+    for (let i = 0; i < errors.length; i++) {
+      const error = errors[i];
+      const element = scopedQuerySelector(this.form_, `[name="${error.name}"]`);
```

Comment:
I got an error in the presubmit on Travis telling me not to `this.form_.querySelector` because of strange behavior. Why is it unnecessary in this case?
If you mean use `document.querySelector`, multiple elements can have identical names so it could select the wrong element.